cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,492 @@
+
+    https://gitlab.com/JoD/exact
+
+    ===============
+    List of classes
+    ===============
+
+    .. autosummary::
+        :nosignatures:
+
+        CPM_exact
+"""
+import sys  # for stdout checking
+import time
+
+from .solver_interface import SolverInterface, SolverStatus, ExitStatus
+from ..expressions.core import *
+from ..expressions.variables import _BoolVarImpl, NegBoolView, _IntVarImpl, _NumVarImpl, intvar
+from ..transformations.comparison import only_numexpr_equality
+from ..transformations.flatten_model import flatten_constraint, flatten_objective, get_or_make_var
+from ..transformations.get_variables import get_variables
+from ..transformations.decompose_global import decompose_global
+from ..transformations.linearize import linearize_constraint, only_positive_bv
+from ..transformations.reification import only_bv_implies, reify_rewrite
+from ..expressions.globalconstraints import DirectConstraint
+import numpy as np
+
+class CPM_exact(SolverInterface):
+    """
+    Interface to the Python interface of Exact
+
+    Requires that the 'exact' python package is installed:
+    $ pip install exact
+    See https://pypi.org/project/exact for more information.
+
+    Creates the following attributes (see parent constructor for more):
+    xct_solver: the Exact instance used in solve()
+    """
+
+    @staticmethod
+    def supported():
+        # try to import the package
+        try:
+            import exact
+            return True
+        except ImportError as e:
+            return False
+
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Requires a CPMpy model as input, and will create the corresponding
+        Exact solver object xct_solver.
+
+        xct_solver can be modified externally before
+        calling solve(), a prime way to use more advanced solver features
+
+        Arguments:
+        - cpm_model: Model(), a CPMpy Model() (optional)
+        """
+        if not self.supported():
+            raise Exception("Install 'exact' as a python package to use this solver interface")
+
+        from exact import Exact as xct
+
+        assert subsolver is None, "Exact does not allow subsolvers."
+
+        # initialise the native solver object
+        self.xct_solver = xct()
+        self.xct_solver.setOption("inp-purelits", "0")    # no dominance breaking to preserve solutions
+        self.xct_solver.setOption("inp-dombreaklim", "0") # no dominance breaking to preserve solutions
+
+        # for solving with assumption variables,
+        self.assumption_dict = None
+
+        # objective can only be set once, so keep track of this
+        self.solver_is_initialized = False
+        self.has_objective = False
+        self.objective_minimize = True
+
+        # initialise everything else and post the constraints/objective
+        super().__init__(name="exact", cpm_model=cpm_model)
+
+    def getSolAndObj(self):
```

Comment:
this looks like a private function, not something you expect a user of the interface to use? If so you could name it `_getSolAndObj()` also it is not really a getter because it returns true/false, so maybe `_fillObjAndVars()`?

Reply:
Done!

Reply:
Done!